### DISCUSSION TITLE
switching events or priority for events

### DISCUSSION THREAD

**Discussion by bary73** [arendst/Tasmota#17154] [Q&A]
I need help setting up for
While event#t1 is running, the temperature for event#t2 Mem7 will reach the set value, event#t1 should turn off for the duration of event#2 and turn on when Mem8 is reached.

Rule1
on DS18B20-1#Temperature do Backlog event t1_demand=%value%;DisplayText [f0s2p15x0y0]Tb. %value% C endon
on DS18B20-2#Temperature do Backlog event t2_demand=%value%;DisplayText [f0s2p15x0y50]Tcwu %value% C endon
on Time#Minute|4 do DisplayText [z] endon
on event#t1_demand>%mem1% DO Backlog dimmer 45;DisplayText [f0s2p15x0y25]Power = 10% break
on event#t1_demand>%mem2% DO Backlog dimmer 59;DisplayText [f0s2p15x0y25]Power = 20% break
on event#t1_demand>%mem3% DO Backlog dimmer 64;DisplayText [f0s2p15x0y25]Power = 30% break
on event#t1_demand>%mem4% DO Backlog dimmer 74;DisplayText [f0s2p15x0y25]Power = 40% break
on event#t1_demand>%mem5% DO Backlog dimmer 84;DisplayText [f0s2p15x0y25]Power = 60% break
on event#t1_demand>%mem6% DO Backlog dimmer 94;DisplayText [f0s2p15x0y25]Power = 80% break
on event#t1_demand<%mem6% DO Backlog dimmer 100;DisplayText [f0s2p15x0y25]Power = MAX endon
on event#t2_demand<%mem7% DO Backlog dimmer 59;DisplayText[z][s2l1c2]CWU = ON break
on event#t2_demand>%mem8% DO rule1 1 endon

mem1 40; mem2 39; mem3 38; mem4 37; mem5 36; mem6 35
mem7 39; mem8 44

**Comment by barbudor** [comment]
I see that you have a `rule1 1` on the last line, do you have another rule with a `rule1 0`
I'm not sure what you call "duration of event" because an event doesn't have a duration, it's a 1 point in time when the event is invoked by the event command.

I feel that you should have 3 rule-sets:
Rule1 that is permanent and never disabled with the first 3 rules
Rule2 with the processing for event_t1 and event_t2 in case XXX and Rule3 with the processing of event_t1 and event_t2 in case YYY
Rule2 and Rule3 being exclusive (so when you eneble one, you disable the other)

So may be something like:
```
Rule1
on DS18B20-1#Temperature do Backlog event t1_demand=%value%;DisplayText [f0s2p15x0y0]Tb. %value% C endon
on DS18B20-2#Temperature do Backlog event t2_demand=%value%;DisplayText [f0s2p15x0y50]Tcwu %value% C endon
on Time#Minute|4 do DisplayText [z] endon

Rule2
on event#t1_demand>%mem1% DO Backlog dimmer 45;DisplayText [f0s2p15x0y25]Power = 10% break
on event#t1_demand>%mem2% DO Backlog dimmer 59;DisplayText [f0s2p15x0y25]Power = 20% break
on event#t1_demand>%mem3% DO Backlog dimmer 64;DisplayText [f0s2p15x0y25]Power = 30% break
on event#t1_demand>%mem4% DO Backlog dimmer 74;DisplayText [f0s2p15x0y25]Power = 40% break
on event#t1_demand>%mem5% DO Backlog dimmer 84;DisplayText [f0s2p15x0y25]Power = 60% break
on event#t1_demand>%mem6% DO Backlog dimmer 94;DisplayText [f0s2p15x0y25]Power = 80% break
on event#t1_demand<%mem6% DO Backlog dimmer 100;DisplayText [f0s2p15x0y25]Power = MAX endon
on event#t2_demand<%mem7% DO Backlog dimmer 59;DisplayText[z][s2l1c2]CWU = ON break
on event#t2_demand>=%mem7% DO Backlog0 Rule2 0; Rule3 1 break

Rule3
on event#t2_demand>%mem8% DO backlog0 rule2 1; rule3 3 endon
```

I say "may be" because I'm not so sure to understand your full use case
You are may be at the point where Rules are no more the solution
Scripting-Language, Berry or backend might be better to implement a state-machine of this complexity

**Comment by bary73** [comment]
`
Rule1 
ON system#boot do Backlog Power8 1;Power4 0 ENDON
ON Power1#state=0 DO Backlog RuleTimer1 120; rule3 0;rule2 0; dimmer 0; DisplayText[z][s2l1c3]STATUS[s3l2c3]OFF;delay 600;DisplayText[z] ENDON 
ON Rules#Timer=1 DO Power2 0 ENDON
ON Power1#state=1 DO Backlog DisplayText[z][s2l1c3]STATUS[s3l2c3]ON;RuleTimer1 0;delay 30;DisplayText[z];rule3 1;rule2 1; delay 5;Power2 1 ENDON
ON Power5#state=1 DO Backlog DisplayText[z][s3l1c2]ALARM!;rule3 0;dimmer 0;delay 60; Power2 0;delay 10; Power1 0 ENDON
ON Power5#state=0 DO DisplayText[z] ENDON
Rule2 
ON Power3#state=1 DO Backlog DisplayText[z][s2l1c2]DEFROST;delay 30; DisplayText[z] ENDON
ON Power3#state=0 DO Backlog DisplayText[z][s3l1c2]WORKING;delay 30; DisplayText[z] ENDON
ON Power4#state=1 DO Backlog DisplayText[z][s2l1c1]NO FLOW ;delay 40;DisplayText[z] ENDON
ON Power4#state=0 DO Backlog DisplayText[z][s3l1c4]OK! ;delay 30;DisplayText[z] ENDON
Rule3 
on DS18B20-1#Temperature do Backlog event temp1_demand=%value%;DisplayText [f0s2p15x0y0]Tb. %value% C endon
on DS18B20-2#Temperature do Backlog event temp2_demand=%value%;DisplayText [f0s2p15x0y50]Cw. %value% C endon
on Time#Minute|4 do DisplayText [z] endon      
on event#temp1_demand>%mem1% DO Backlog dimmer 45;DisplayText [f0s2p15x0y25]POWER = 10% break
on event#temp1_demand>%mem2% DO Backlog dimmer 59;DisplayText [f0s2p15x0y25]POWER = 20% break    
on event#temp1_demand>%mem3% DO Backlog dimmer 64;DisplayText [f0s2p15x0y25]POWER = 30% break   
on event#temp1_demand>%mem4% DO Backlog dimmer 73;DisplayText [f0s2p15x0y25]POWER = 40% break     
on event#temp1_demand>%mem5% DO Backlog dimmer 84;DisplayText [f0s2p15x0y25]POWER = 60% break  
on event#temp1_demand>%mem6% DO Backlog dimmer 94;DisplayText [f0s2p15x0y25]POWER = 80% break   
on event#temp1_demand<%mem6% DO Backlog dimmer 100;DisplayText [f0s2p15x0y25]POWER = MAX endon

on event#temp2_demand<%mem7% DO Backlog Power6 1;dimmer 59;DisplayText[z][f0s2p15x0y0y0]DHW operation;DisplayText [f0s2p15x0y50]Cw. %value% C  break
on event#temp2_demand>%mem8% DO Power6 0 endon
`
This is my current code where I wanted to prioritize event#temp2_demand over event#temp1_demand
